packages/aws-rfdk/lib/deadline/lib/thinkbox-docker-images.ts

```diff
@@ -0,0 +1,224 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { randomBytes } from 'crypto';
+import * as path from 'path';
+
+import {
+  Repository,
+} from '@aws-cdk/aws-ecr';
+import {
+  ContainerImage,
+} from '@aws-cdk/aws-ecs';
+import {
+  Code,
+  SingletonFunction,
+  Runtime,
+} from '@aws-cdk/aws-lambda';
+import { RetentionDays } from '@aws-cdk/aws-logs';
+import {
+  Construct,
+  CustomResource,
+  Duration,
+  Token,
+} from '@aws-cdk/core';
+
+import {
+  IVersion,
+  RenderQueueImages,
+  ThinkboxManagedDeadlineDockerRecipes,
+  UsageBasedLicensingImages,
+} from '.';
+
+/**
+ * Interface to specify the properties when instantiating a {@link ThinkboxDockerImages} instnace.
+ */
+export interface ThinkboxDockerImagesProps {
+  /**
+   * An optional region for the ECR repository.
+   * @default global
+   */
+  readonly region?: string;
+
+  /**
+   * The Deadline version to obtain images for.
+   * @default latest
+   */
+  readonly version?: IVersion
+}
+
+/**
+ * An API for interacting with staged Deadline Docker images provided by AWS Thinkbox.
+ *
+ * This provides container images as required by RFDK's Deadline constructs such as
+ *
+ * * {@link @aws-rfdk/deadline#RenderQueue}
+ * * {@link @aws-rfdk/deadline#UsageBasedLicensing}
+ *
+ * @example Construct a RenderQueue
+ *
+ * import { App, Stack, Vpc } from '@aws-rfdk/core';
+ * import { RenderQueue, Repository, ThinkboxDockerRecipes } from '@aws-rfdk/deadline';
+ * const app = new App();
+ * const stack = new Stack(app, 'Stack');
+ * const vpc = new Vpc(app, stack);
+ * const images = new ThinkboxDockerImages(stack, 'Image'});
+ * const repository = new Repository(stack, 'Repository', {
+ *   vpc,
+ *   version: recipes.version
+ * });
+ *
+ * const renderQueue = new RenderQueue(stack, 'RenderQueue', {
+ *   images: images.forRenderQueue(),
+ *   // ...
+ * });
+ */
+export class ThinkboxDockerImages extends Construct {
+  /**
+   * A {@link DockerImageAsset} that can be used to build Thinkbox's Deadline RCS Docker Recipe into a
+   * container image that can be deployed in CDK.
+   *
+   * @param scope The parent scope
+   * @param id The construct ID
+   */
+  public readonly remoteConnectionServer: ContainerImage;
+
+  /**
+   * A {@link DockerImageAsset} that can be used to build Thinkbox's Deadline License Forwarder Docker Recipe into a
+   * container image that can be deployed in CDK.
+   *
+   * @param scope The parent scope
+   * @param id The construct ID
+   */
+  public readonly licenseForwarder: ContainerImage;
+
+  /**
+   * The version of Deadline in the stage directory.
+   */
+  public readonly version?: IVersion;
+
+  constructor(scope: Construct, id: string, props?: ThinkboxDockerImagesProps) {
+    super(scope, id);
+
+    this.version = props?.version;
+
+    const lambdaCode = Code.fromAsset(path.join(__dirname, '..', '..', 'lambdas', 'nodejs'));
+
+    const lambdaFunc = new SingletonFunction(this, 'VersionProviderFunction', {
```

Comment:
AFAIK, there isn't any difference between just a `Function` and a `SingletonFunction` in this particular use. Anchoring the `SingletonFunction` on the parent stack, instead, allows it to truly be a singleton that's unique to the stack.

Reply:
According to the [CDK `SingletonFunction` docs](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-lambda-readme.html#singleton-function), as long as the `uuid` and the (optional) `lambdaPurpose` properties are the same within a stack, this is guaranteed to be a singleton for that stack (see [source code](https://github.com/aws/aws-cdk/blob/d613447c3161d698cce739fbcdd02042ad65b9bd/packages/%40aws-cdk/aws-lambda/lib/singleton-lambda.ts#L111-L121)).

Reply:
Confirmed. Resolving